### PR TITLE
feat: per-position activation patching + causal metrics (roadmap Tier 1 #1)

### PR DESCRIPTION
**Pull request by edward-lcl**
Implements the bulk of `CAUSAL_INTERP_ROADMAP.md` Tier 1 item 1 (Activation Patching API). Five of six checkboxes ticked; the remaining one (component types `resid_pre`/`resid_post`/`attn_head`) is deferred because it needs deeper component-level architectural introspection than the current MLP/attn/output split exposes.

This is independent of #10 (Text Generation with Interventions) and can land separately.

## What's new

```python
# Existing layer-only behaviour preserved (returns Dict[int, float])
results = model.activation_patching(
    clean_text="Paris is the capital of France",
    corrupted_text="London is the capital of France",
    component="output",
)

# NEW: per-position patching, returns (n_layers, n_positions) numpy matrix
matrix = model.activation_patching(
    clean_text="Paris is the capital of France",
    corrupted_text="London is the capital of France",
    component="output",
    positions="all",
    metric="logit_diff",
    clean_target=" Paris",
    corrupted_target=" London",
    plot=True,
)
```

### `positions` parameter

- `None` (default): unchanged behaviour, returns `Dict[int, float]`
- list of ints or `"all"`: patches only those positions in the corrupted run, returns `(n_layers, n_positions)` numpy matrix

Position-level patching requires `clean_text` and `corrupted_text` to tokenise to the same length. The method validates this and raises a clear `ValueError` otherwise.

### Causal metrics (built-ins)

- `logit_diff`: requires `clean_target` and `corrupted_target`. Measures how much of the clean run's logit advantage on the target token the patch recovers. Standard mech-interp metric.
- `kl_divergence`: KL(softmax(clean) || softmax(patched)). Lower = closer to clean.
- `cross_entropy_diff`: CE(clean_dist, patched) − CE(clean_dist, clean).
- existing `l2`, `cosine`, `mse` unchanged
- arbitrary callables: pass a `metric=lambda patched, clean, corrupted: ...` for custom scoring

### Heatmap visualization

`plot=True` with `positions != None` renders a symmetric `RdBu_r` heatmap (positions on x-axis, layers on y-axis). Existing layer-only bar plot path unchanged.

### Supporting helper: `interventions.replace_at_positions(value, positions)`

Overwrites only the specified token positions in an activation, leaving the rest intact. Built with `mx.concatenate` over slices (no in-place mutation). Used internally by the matrix path; exposed publicly because it's generally useful for any per-position intervention.

## Validation

Tested on `lmstudio-community/Qwen3-4B-Instruct-2507-MLX-8bit` (M5 Pro, MLX 0.29):

- Layer-only call: unchanged dict return, all 6 layers tested at 100% recovery on the canonical Paris/London task
- `positions='all'`, l2 metric: 6 layers × 6 positions matrix in 1.8s, mean 14.4% recovery, max 90.5% — sensible per-position distribution
- `positions=[0, 3, 5]`, logit_diff metric: matrix shape (6, 3); late layers at later positions show the expected stronger recovery (+25.2% at layer 30, position 5)
- Same-length validation: raises ValueError on token-length mismatch with a message pointing the user at positions=None or padding

Each forward is ~50ms; a 36-cell patching grid runs in ~2 seconds. Existing test suite passes.

## Files changed

- `mlxterp/analysis.py` (+340): `activation_patching` extended with `positions`, `clean_target`, `corrupted_target`; new private `_activation_patching_matrix()` and `_plot_patching_matrix()`
- `mlxterp/core/intervention.py` (+83): `replace_at_positions()` helper
- `mlxterp/core/__init__.py` (+2): export new helper
- `mlxterp/__init__.py` (+3): expose in `interventions` namespace
- `CAUSAL_INTERP_ROADMAP.md` (+27, -7): tick checkboxes with implementation notes
- `tests/test_activation_patching_positions.py` (+139): 8 contract tests (force-added per the repo's `test_*.py` gitignore convention)
- `examples/activation_patching_positions.py` (+82): canonical Paris/London demo

## Notes for review

- **Component types `resid_pre`/`resid_post`/`attn_head`** are intentionally deferred. The current code supports `mlp`, `self_attn`, `output` (the full layer output, which is already roughly `resid_post`). True residual-stream-pre and per-attention-head patching needs the proxy/wrapping system to expose those components individually; happy to follow up in a separate PR if you'd prefer.
- **`logit_diff` target tokenisation** uses `tokenizer.encode(target)[0]` (first token). Most space-prefixed targets like `" Paris"` tokenise to a single token in MLX-LM models so this works as-is, but a more robust version would either accept token IDs directly or warn on multi-token targets. Open to feedback.
- **Same-length input constraint** for position-level patching is a hard requirement under the current implementation. End-aligning would be the natural extension; not included here to keep the PR focused.
- **Test file** force-added because the repo's `.gitignore` has `test_*.py` (same approach as #10).

Built and validated as part of [AJAR Hidden CoT detection](https://github.com/edward-lcl/AJAR). Validation harness: [`scripts/mlxterp_patching_validate.py`](https://github.com/edward-lcl/AJAR/blob/main/scripts/mlxterp_patching_validate.py).

### CODE REVIEW

**Comment by edward-lcl** [comment]
Same 5/5 parity result applies to this PR too — see comment on #10. The activation_patching extensions in this PR are independently validated in [`scripts/mlxterp_patching_validate.py`](https://github.com/edward-lcl/AJAR/blob/main/scripts/mlxterp_patching_validate.py) (8/8 contract checks passed, including the logit_diff metric on the canonical Paris/London task showing the expected late-layer/late-position causal gradient).
